File: modules/digital-carbon-usage-bridge/ios/DigitalCarbonUsageBridgeModule.swift
import ExpoModulesCore

public class DigitalCarbonUsageBridgeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DigitalCarbonUsageBridge")

    Function("getBridgeStatus") {
      return [
        "moduleName": "DigitalCarbonUsageBridge",
        "platform": "ios",
        "installed": true,
        "supportsDeviceWideUsage": false,
        "accessGranted": false,
        "requiresManualAccess": false,
        "canOpenSettings": false,
        "note": "iOS does not expose device-wide app usage history to this app surface, so the app stays on local fallbacks here.",
        "supportedMetrics": []
      ]
    }

    AsyncFunction("getTodayUsageSnapshot") { () -> [String: Any]? in
      return nil
    }

    AsyncFunction("openUsageAccessSettings") { () -> Bool in
      return false
    }
  }
}

File: modules/digital-carbon-usage-bridge/src/DigitalCarbonUsageBridgeModule.ts
import { requireOptionalNativeModule } from 'expo';
import { Platform } from 'react-native';

import type {
  DigitalCarbonUsageBridgeModuleType,
  DigitalCarbonUsageBridgeStatus,
  DigitalCarbonUsageSnapshot,
} from './DigitalCarbonUsageBridge.types';

const fallbackStatus: DigitalCarbonUsageBridgeStatus = {
  moduleName: 'DigitalCarbonUsageBridge',
  platform: Platform.OS,
  installed: false,
  supportsDeviceWideUsage: false,
  accessGranted: false,
  requiresManualAccess: Platform.OS === 'android',
  canOpenSettings: false,
  note:
    Platform.OS === 'android'
      ? 'Native app usage bridge is not built into this binary yet.'
      : 'This platform is still using local fallbacks for app usage.',
  supportedMetrics: [],
};

const fallbackModule: DigitalCarbonUsageBridgeModuleType = {
  getBridgeStatus() {
    return fallbackStatus;
  },
  async getTodayUsageSnapshot(): Promise<DigitalCarbonUsageSnapshot | null> {
    return null;
  },
  async openUsageAccessSettings() {
    return false;
  },
};

export default (
  requireOptionalNativeModule<DigitalCarbonUsageBridgeModuleType>(
    'DigitalCarbonUsageBridge',
  ) ?? fallbackModule
);
